From 895157b061558713c9ed7df95a4b328c137949fe Mon Sep 17 00:00:00 2001
From: John Andersen <john.s.andersen@intel.com>
Date: Mon, 14 Sep 2020 22:43:40 -0700
Subject: [PATCH 06/11] X86: cpu: Improve clarity of pinned CR0 WP bit

In order to improve clarity of paravirtualized control register pinning
using the same bits as native pinning, introduce a cr0_pinned_bits const
which is analogous to cr4_pinned_bits.

Signed-off-by: John Andersen <john.s.andersen@intel.com>
---
 arch/x86/kernel/cpu/common.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/arch/x86/kernel/cpu/common.c b/arch/x86/kernel/cpu/common.c
index 002a48abf777..30db3873f3c5 100644
--- a/arch/x86/kernel/cpu/common.c
+++ b/arch/x86/kernel/cpu/common.c
@@ -347,6 +347,9 @@ static __always_inline void setup_umip(struct cpuinfo_x86 *c)
 	cr4_clear_bits(X86_CR4_UMIP);
 }
 
+#define CR0_PINNED_MASK X86_CR0_WP
+static const unsigned long cr0_pinned_mask = CR0_PINNED_MASK;
+static const unsigned long cr0_pinned_bits = CR0_PINNED_MASK;
 /* These bits should not change their value after CPU init is finished. */
 static const unsigned long cr4_pinned_mask =
 	X86_CR4_SMEP | X86_CR4_SMAP | X86_CR4_UMIP | X86_CR4_FSGSBASE;
@@ -355,19 +358,20 @@ static unsigned long cr4_pinned_bits __ro_after_init;
 
 void native_write_cr0(unsigned long val)
 {
-	unsigned long bits_missing = 0;
+	unsigned long bits_changed = 0;
 
 set_register:
 	asm volatile("mov %0,%%cr0": "+r" (val), "+m" (__force_order));
 
 	if (static_branch_likely(&cr_pinning)) {
-		if (unlikely((val & X86_CR0_WP) != X86_CR0_WP)) {
-			bits_missing = X86_CR0_WP;
-			val |= bits_missing;
+		if (unlikely((val & cr0_pinned_mask) != cr0_pinned_bits)) {
+			bits_changed = (val & cr0_pinned_mask) ^ cr0_pinned_bits;
+			val = (val & ~cr0_pinned_mask) | cr0_pinned_bits;
 			goto set_register;
 		}
-		/* Warn after we've set the missing bits. */
-		WARN_ONCE(bits_missing, "CR0 WP bit went missing!?\n");
+		/* Warn after we've corrected the changed bits. */
+		WARN_ONCE(bits_changed, "pinned CR0 bits changed: 0x%lx!?\n",
+			  bits_changed);
 	}
 }
 EXPORT_SYMBOL(native_write_cr0);
-- 
2.21.0

